tensorflow/core/kernels/tensor_to_hash_bucket_op_gpu.cu.cc

```diff
@@ -0,0 +1,111 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#if GOOGLE_CUDA
+
+#define EIGEN_USE_GPU
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/kernels/tensor_to_hash_bucket_op.h"
+#include "third_party/farmhash_gpu/src/farmhash_gpu.h"
+#include "tensorflow/core/util/gpu_kernel_helper.h"
+
+namespace tensorflow {
+
+namespace {
+
+constexpr int BUFFER_SIZE = 64;
+
+template<typename T>
+__device__ void FillDigits(T val, int num_digits, int* i, char *buf) {
+  eigen_assert(num_digits <= BUFFER_SIZE - (*i));
+
+  int factor = (val < 0 ? -1: 1);
+
+  int num_digits_a = num_digits;
+  while (num_digits > 0) {
+    int digit = static_cast<int>((val % 10) * factor);
+    buf[(*i) + num_digits - 1] = digit + '0';
+    val /= 10;
+    num_digits--;
+  }
+
+  (*i) += num_digits_a;
+}
+
+template<typename T>
+__device__ int IntegerToString(T val, char *buf) {
+  int num_digits = 0;
+  T val_a = val;
+  do {
+    val_a = val_a / 10;
+    num_digits++;
+  } while(val_a != 0);
```

Comment:
This is borderline nitpicky, but can't we populate `buf` in this loop itself and then `std::reverse` it?

Reply:
I think we cannot use `std::reverse` in the device function, right?

Reply:
Can you check?  It should compile down to a loop, so I don't immediately see why we can't use it.

Reply:
Do you mean call the host side function `std::reverse` in the device function? It couldn't build can gave 
```
error: calling a __host__ function("std::reverse<char *> ") from a __device__ function("tensorflow::_NV_ANON_NAMESPACE::IntegerToString<short> ") is not allowed
````

Reply:
But let me try to merge the two loops.

Reply:
I feel we still need to use two loops for this. Considering filling the digits is naturally from the backward direction, we can (1) compute the num_digits and then fill the digits backward (as the current code) or (2) fill it forward and reverse it later which still needs two loops.